lora: asynchronous packet reception

JordanYates opened the pull request:
Extends the LoRa API to enable asynchronous packet reception.

In many situations applications aren't waiting for a packet in particular, but just want to handle any packet that comes in.
Using the asynchronous API allows us to avoid the creation of a dedicated thread to continuously block on `lora_recv`.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
samples: lora: receive: demo async reception

Demonstrate asynchronous reception in the LoRa sample.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Mani-Sadhasivam commented on drivers/lora/sx12xx_common.c:
```diff
@@ -144,6 +163,33 @@ int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
+	}
+
+	/* Handle reception already running */
+	if (dev_data.rx_cb) {
```
This is going to be problematic :( What if both the ISR and the application tries to modify the callback? We are in a race, right? And do you really have a use case to switch the callbacks in runtime? Otherwise, please get rid of it.

JordanYates replied:
I don't have a use case for it, but thought someone might, some sort of packet state machine perhaps.

Yes this is technically a race condition, but so is every other function in this driver, none of them will work without external synchronization across multiple threads.

The idea was that once started, either the initiating thread OR the callback is responsible for future calls to `lora_recv_async`, but not both of them arbitrarily. As this is basically the same requirement as "2 threads cannot call `lora_recv` at the same time", I assumed it wouldn't be a problem.

Mani-Sadhasivam replied:
> Yes this is technically a race condition, but so is every other function in this driver, none of them will work without external synchronization across multiple threads.

Right, the reason was the use cases were simple at that point. But I'm pretty sure that we need to rewrite the driver in the future to handle concurrency when developers report issues or someone has time to look into it. But here, since the callback assignment is shared between the ISR and userspace, the issue is obvious. So I think you can just remove the callback reassigning part for now.

Commit:
lora: asynchronous packet reception

Adds functionality to receive LoRa packets asynchronously. Reception
runs continuously until cancelled by another call to `lora_recv_async`.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
@Mani-Sadhasivam @mniestroj This is now ready for review

Mani-Sadhasivam reviewed (approved):
LGTM
